core/whistle_apps-1.0.0/src/whapps_maintenance.erl

```diff
@@ -279,6 +294,144 @@ maybe_remove_invalid_notify_doc(_Type, _Id, _Doc) -> 'ok'.
 %%--------------------------------------------------------------------
 %% @public
 %% @doc
+%% This function will move a system config setting from one location
+%%  to another.  It will create the document if it does not already
+%%  exists and will move per-node settings if they exist.  It can
+%%  handle moving complex or nested keys.
+%%  In the event that both the source and destination exist but
+%%  have different values it will not make any change.  The parameter
+%%  is only removed from the source after a successsful save of the
+%%  the destiation.
+%% @end
+%%--------------------------------------------------------------------
+-type migrate_setting() :: {ne_binary(), ne_binary()}.
+-type migrate_value() :: {ne_binary(), ne_binary(), ne_binary() | ne_binaries(), _}.
+-type migrate_values() :: [migrate_value()].
+-spec migrate_config_setting(migrate_setting(), migrate_setting()) ->
+                                    'ok' |
+                                    {'error', _}.
+migrate_config_setting(From, To) ->
+    case remove_config_setting(From) of
+        {'ok', _, []} -> 'ok';
+        {'ok', JObj, Removed} ->
+            migrate_config_setting(JObj, Removed, To);
+        {'error', 'not_found'} -> 'ok';
+        {'error', Reason} -> {'error', {'remove', Reason}}
+    end.
+
+-spec migrate_config_setting(wh_json:object(), migrate_values(), migrate_setting()) ->
+                                    'ok' |
+                                    {'error', _}.
+migrate_config_setting(UpdatedFrom, Removed, To) ->
+    case add_config_setting(To, Removed) of
+        {'ok', UpdatedTo} ->
+            {'ok', _} = couch_mgr:save_doc(?WH_CONFIG_DB, UpdatedTo),
```

Comment:
@jamesaimonetti If the first fails no change has been made.  If the second fails then the first has stored the value in the new location and simply failed to remove the old location.  In this state the value exists in both places so it is available to the new code, the only downside is the old value is orphaned.  IMHO this is better than reverting as that would mean the value at the new location would be written with the default (since it was missing) if requested prior to admin resolving it, and that would likely change system behavior.